server/src/test/java/org/elasticsearch/indices/state/CloseWhileRelocatingShardsIT.java

```diff
@@ -0,0 +1,208 @@
+
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.cluster.routing.allocation.decider.ConcurrentRebalanceAllocationDecider.CLUSTER_ROUTING_ALLOCATION_CLUSTER_CONCURRENT_REBALANCE_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.decider.EnableAllocationDecider.CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.decider.ThrottlingAllocationDecider.CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING;
+import static org.elasticsearch.indices.state.CloseIndexIT.assertException;
+import static org.elasticsearch.indices.state.CloseIndexIT.assertIndexIsClosed;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+
+@ESIntegTestCase.ClusterScope(minNumDataNodes = 2)
+public class CloseWhileRelocatingShardsIT extends ESIntegTestCase {
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+            .put(super.nodeSettings(nodeOrdinal))
+            .put(CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey(), 10)
+            .put(CLUSTER_ROUTING_ALLOCATION_CLUSTER_CONCURRENT_REBALANCE_SETTING.getKey(), -1)
+            .build();
+    }
+
+    @Override
+    protected int numberOfReplicas() {
+        return 1;
+    }
+
+    public void testCloseWhileRelocatingShards() throws Exception {
+        final String[] indices = new String[randomIntBetween(3, 10)];
+        final Map<String, AtomicInteger> docsPerIndex = new HashMap<>();
+
+        for (int i = 0; i < indices.length; i++) {
+            final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            createIndex(indexName);
+
+            int nbDocs = 0;
+            if (randomBoolean()) {
+                nbDocs = randomIntBetween(1, 20);
+                indexRandom(randomBoolean(), false, randomBoolean(), IntStream.range(0, nbDocs)
+                    .mapToObj(n -> client().prepareIndex(indexName, "_doc", String.valueOf(n)).setSource("num", n)).collect(toList()));
+                docsPerIndex.put(indexName, new AtomicInteger(nbDocs));
+            }
+            indices[i] = indexName;
+            docsPerIndex.put(indexName, new AtomicInteger(nbDocs));
+        }
+
+        ensureGreen(indices);
+        assertAcked(client().admin().cluster().prepareUpdateSettings()
+            .setTransientSettings(Settings.builder()
+                .put(CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING.getKey(), EnableAllocationDecider.Rebalance.NONE.toString())));
+
+        final List<Thread> indexingThreads = new ArrayList<>();
+        final AtomicBoolean indexing = new AtomicBoolean(true);
+
+        // start some concurrent indexing threads
+        for (final String index : indices) {
+            if (randomBoolean()) {
+                final Thread thread = new Thread(() -> {
```

Comment:
there's a utility class called BackgroundIndexer which might be of help here, see `testRelocationWhileIndexingRandom`.

Reply:
Thanks, I didn't know about this one.